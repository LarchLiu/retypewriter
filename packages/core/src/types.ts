export interface InsertPatch {
  type: 'insert'
  cursor: number
  content: string
}

export interface RemovalPatch {
  type: 'removal'
  cursor: number
  length: number
}

export type Patch = InsertPatch | RemovalPatch

export interface Slice {
  content: string
  order: number
  cursor: number
}

export interface Snapshot {
  content: string
  options?: SnapshotOptions
}

export interface SnapshotOptions {
  wait?: number
  pause?: boolean
}

export interface AnimatorStepInsert {
  type: 'insert' | 'delete'
  cursor: number
  content: string
  char: string
}

export interface AnimatorStepRemoval {
  type: 'removal'
  cursor: number
  content: string
}

export interface AnimatorStepInit {
  type: 'init'
  content: string
}

export interface AnimatorStepPatch {
  type: 'new-patch'
  patch: Patch
  index: number
}

export interface AnimatorStepSnap {
  type: 'new-snap'
  snap: Snapshot
  index: number
}

export interface AnimatorStepSnapFinish {
  type: 'snap-finish'
  content: string
}

export type AnimatorStep =
  | AnimatorStepInsert
  | AnimatorStepRemoval
  | AnimatorStepInit
  | AnimatorStepPatch
  | AnimatorStepSnap
  | AnimatorStepSnapFinish

export interface ParsedSnaphot {
  raw: string
  start: number
  end: number
  body: string
  bodyStart: number
  bodyEnd: number
  optionsRaw?: string
  options?: SnapshotOptions
}

export interface ParsedHead {
  options?: SnapshotOptions
}
