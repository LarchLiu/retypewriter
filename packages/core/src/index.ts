export * from './steps'
export * from './types'
export * from './patch'
export * from './snaps'
export * from './slicing'
export * from './timing'
export * from './typewriter'
