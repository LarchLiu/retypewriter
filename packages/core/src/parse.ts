import YAML from 'js-yaml'
import type { ParsedHead, ParsedSnaphot, Snapshot, SnapshotOptions } from './types'

export const SNAP_EXT = '.retypewriter'
export const SNAP_HEADING = 'reTypewriter Snapshots v1\n'
export const SNAP_SEPERATOR_PRE = '-'.repeat(2)
export const SNAP_SEPERATOR_POST = '-'.repeat(10)
export const SNAP_SEPERATOR = `${SNAP_SEPERATOR_PRE}--${SNAP_SEPERATOR_POST}`
export const SNAP_SEPERATOR_OPTIONS = '-----options--'
export const SNAP_SEPERATOR_MATCHER = new RegExp(`\\n?${SNAP_SEPERATOR_PRE}[\\w-]{2}${SNAP_SEPERATOR_POST}\\n`, 'g')
export const SNAP_SEPERATOR_OPTIONS_MATCHER = new RegExp(`\\n?${SNAP_SEPERATOR_OPTIONS}\\n`, 'g')

export function parseOptions(raw: string): SnapshotOptions | undefined {
  raw = raw.trim()
  if (!raw)
    return undefined
  return (raw.startsWith('{'))
    ? JSON.parse(raw)
    : YAML.load(raw)
}

export function stringifySnapshots(snapshots: Snapshot[], useYaml = true) {
  return [
    SNAP_HEADING,
    ...snapshots
      .flatMap((snap, i) => [
        SNAP_SEPERATOR_PRE + (i + 1).toString().padStart(2, '0') + SNAP_SEPERATOR_POST,
        snap.content,
        ...(
          snap.options
            ? [
              SNAP_SEPERATOR_OPTIONS,
              useYaml
                ? YAML.dump(snap.options, { indent: 2 }).trimEnd()
                : Object.keys(snap.options).length > 1
                  ? JSON.stringify(snap.options, null, 2)
                  : JSON.stringify(snap.options),
            ]
            : []
        ),
      ]),
    SNAP_SEPERATOR,
    '',
  ].join('\n')
}

export function parseSnapshots(raw: string) {
  if (!raw.startsWith(SNAP_HEADING))
    throw new SyntaxError('Invalid snapshot file')
  if (!raw.endsWith('\n'))
    raw += '\n'

  const head: ParsedHead = {}
  const snaps: ParsedSnaphot[] = []

  SNAP_SEPERATOR_MATCHER.lastIndex = 0
  const matches = Array.from(raw.matchAll(SNAP_SEPERATOR_MATCHER))

  matches.forEach((match, idx) => {
    const start = match.index!

    if (idx === 0) {
      const content = raw.slice(0, start)
      head.options = parseOptions(content.slice(SNAP_HEADING.length))
    }

    const next = matches[idx + 1]
    const end = next?.index ?? raw.length
    const bodyStart = start + match[0].length
    const withOptions = raw.slice(bodyStart, end)
    const parts = withOptions.split(SNAP_SEPERATOR_OPTIONS_MATCHER)
    const body = parts[0]
    const bodyEnd = bodyStart + body.length
    const snap: ParsedSnaphot = {
      raw: raw.slice(start, end),
      start,
      end,
      body,
      bodyEnd,
      bodyStart,
    }
    const optionsRaw = parts[1]?.trim()
    if (optionsRaw) {
      snap.optionsRaw = optionsRaw
      snap.options = parseOptions(optionsRaw)
    }
    snaps.push(snap)
  })

  // remove tailing separator
  snaps.pop()

  return {
    head,
    snapshots: snaps,
  }
}
