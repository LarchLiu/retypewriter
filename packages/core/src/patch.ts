import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'
import type { Patch } from './types'

export function diff(a: string, b: string): Diff[] {
  const differ = new DMP()
  const delta = differ.diff_main(a, b)
  differ.diff_cleanupSemantic(delta)
  return delta
}

export function calculatePatch(diff: Diff[]): Patch[] {
  const patches: Patch[] = []

  let cursor = 0
  for (const change of diff) {
    if (change[0] === 0) {
      cursor += change[1].length
    }
    else if (change[0] === -1) {
      const length = change[1].length
      patches.push({
        type: 'removal',
        cursor: cursor + length,
        length,
      })
    }
    else if (change[0] === 1) {
      const content = change[1]
      const tailingNewLines = content.match(/\n*$/m)?.[0] || ''
      const rest = tailingNewLines.length
        ? content.slice(0, -tailingNewLines.length)
        : content
      // if there is new lines in the patch, we should type them first
      if (tailingNewLines.length) {
        patches.push({
          type: 'insert',
          cursor,
          content: tailingNewLines,
        })
      }
      if (rest.length) {
        patches.push({
          type: 'insert',
          cursor,
          content: rest,
        })
      }
      cursor += content.length
    }
    else {
      throw new Error('unknown change type')
    }
  }
  return patches
}
