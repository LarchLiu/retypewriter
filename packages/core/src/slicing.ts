import type { Slice } from './types'

const pairs = [
  ['{', '}'],
  ['(', ')'],
  ['[', ']'],
  ['<', '>'],
  ['`', '`'],
  ['\'', '\''],
  ['"', '"'],
]

const chars = '(){}\\[\\]<>\'"`'
const splitRegex = new RegExp(`(\\s*[${chars}]\\s*)`)

export function sliceInput(input: string): Slice[] {
  if (!input)
    return []
  const tailingNewLines = input.match(/\n+$/m)?.[0] || ''
  const rest = tailingNewLines.length
    ? input.slice(0, -tailingNewLines.length)
    : input

  const strings = rest.split(splitRegex).filter(Boolean)

  let index = 0
  const slices = strings
    .map((s, idx) => {
      const trimmed = s.trim()
      const pair = pairs.find(i => i[1] === trimmed)
      const order = (pair && strings[idx - 2]?.trim() === pair[0])
        ? idx - 2 // paired with previous
        : idx

      const _index = index
      index += s.length

      return {
        content: s,
        order,
        cursor: _index,
      }
    })
    .sort((a, b) => a.order - b.order)

  let cursor = 0
  slices.forEach((i) => {
    i.cursor = Math.min(i.cursor, cursor)
    cursor += i.content.length
  })

  if (tailingNewLines.length) {
    slices.unshift({
      content: tailingNewLines,
      order: -1,
      cursor: 0,
    })
  }

  return slices
}

export function applySlice(slices: Slice[]) {
  const result = slices.reduce((acc, i) => {
    const { content: contet, cursor } = i
    return acc.slice(0, cursor) + contet + acc.slice(cursor)
  }, '')
  return result
}
