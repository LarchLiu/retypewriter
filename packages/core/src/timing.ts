export const timeRanges: [string, number, number][] = [
  // default
  ['', 0, 100],
  // commons
  [' ', 0, 40],
  ['adefijkloosu', 0, 50],
  ['bcghmnpqrtvwxyz', 5, 70],
  // uppercase
  ['adefijkloosu'.toUpperCase(), 0, 80],
  ['bcghmnpqrtvwxyz'.toUpperCase(), 5, 100],
  // a bit far
  ['[];\',./-=', 10, 120],
  // need shift
  ['<>?:"{}|\\_+()*', 20, 140],
  // a bit far and shift
  ['!~`#$%^&', 20, 180],
]

export function randRange(min: number, max: number): number {
  return Math.random() * (max - min) + min
}

export function getTimeout(char: string, multiplier = 1): number {
  const [, min, max] = timeRanges.find(range => range[0].includes(char)) || timeRanges[0]
  return randRange(min, max) * multiplier
}
