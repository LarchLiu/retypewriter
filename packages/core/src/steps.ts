import { calculatePatch, diff } from './patch'
import { sliceInput } from './slicing'
import type { AnimatorStep, Patch, Snapshot } from './types'

export function *patchSteps(input: string, patches: Patch[]): Generator<AnimatorStep> {
  let output = input
  let cursor = 0

  for (let index = 0; index < patches.length; index++) {
    const patch = patches[index]

    yield {
      type: 'patch-start',
      patch,
      index,
      total: patches.length,
    }

    if (patch.type === 'insert') {
      cursor = patch.cursor
      const head = output.slice(0, patch.cursor)
      const tail = output.slice(patch.cursor)
      for (const { char, output, cursor: delta } of animateInsertionSlices(patch.content)) {
        yield {
          type: 'insert',
          char,
          cursor: cursor + delta,
          content: head + output + tail,
        }
      }
      output = head + patch.content + tail
    }
    else if (patch.type === 'removal') {
      cursor = patch.cursor - patch.length
      const head = output.slice(0, cursor)
      const tail = output.slice(patch.cursor)
      const selection = output.slice(cursor, patch.cursor)
      for (let i = selection.length - 1; i >= 0; i--) {
        yield {
          type: 'removal',
          cursor: cursor + i,
          content: head + selection.slice(0, i) + tail,
        }
      }
      output = head + tail
    }

    yield {
      type: 'patch-finish',
      content: output,
      index,
      total: patches.length,
    }
  }
}

export function *animateSteps(snapshots: Snapshot[]): Generator<AnimatorStep> {
  let lastContent: string | undefined
  const copy = [...snapshots]
  for (let index = 0; index < copy.length; index++) {
    const snap = copy[index]
    if (lastContent == null) {
      lastContent = snap.content
      yield {
        type: 'init',
        content: lastContent,
      }
      continue
    }

    yield {
      type: 'snap-start',
      snap,
      index,
      total: copy.length,
    }

    const steps = stepsTo(lastContent, snap.content)
    for (const step of steps)
      yield step

    yield {
      type: 'snap-finish',
      content: snap.content,
      index,
      total: copy.length,
    }

    lastContent = snap.content
  }
}

export function *animateInsertionSlices(input: string) {
  const slices = sliceInput(input)
  let output = ''
  for (const { content, cursor } of slices) {
    const head = output.slice(0, cursor)
    const tail = output.slice(cursor)

    let body = ''
    for (const char of content) {
      body += char
      yield {
        char,
        output: head + body + tail,
        cursor: cursor + body.length,
      }
    }
    output = head + content + tail
  }
}

export function stepsTo(input: string, output: string) {
  const delta = diff(input, output)
  const patches = calculatePatch(delta)
  return patchSteps(input, patches)
}

export function applyPatches(input: string, patches: Patch[]) {
  let output = input
  for (const patch of patchSteps(input, patches)) {
    if (patch.type === 'patch-finish')
      output = patch.content
  }
  return output
}
