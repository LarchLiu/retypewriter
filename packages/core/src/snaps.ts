import type { AnimatorStep, Snapshot } from './types'
import { stepsTo } from './steps'
import type { TypewriterOptions } from './typewriter'
import { typingAnimator } from './typewriter'
import { SNAP_EXT, parseSnapshots, stringifySnapshots } from './parse'

export class Snapshots extends Array<Snapshot> {
  constructor(...args: Snapshot[]) {
    super(...args)
  }

  get last() {
    return this[this.length - 1]
  }

  get first() {
    return this[0]
  }

  move(from: number, to: number) {
    if (from === to)
      return

    const element = this[from]
    this.splice(from, 1)
    this.splice(to, 0, element)
  }

  toString(useYaml = true) {
    return stringifySnapshots(this, useYaml)
  }

  fromString(raw: string) {
    const { snapshots: parsed } = parseSnapshots(raw)
    this.length = 0
    parsed.forEach((p) => {
      this.push({
        content: p.body,
        options: p.options,
      })
    })
    return this
  }

  static fromString(raw: string) {
    return new Snapshots().fromString(raw)
  }

  *steps(): Generator<AnimatorStep> {
    let lastContent: string | undefined
    const copy = [...this]
    for (let index = 0; index < copy.length; index++) {
      const snap = copy[index]
      if (lastContent == null) {
        lastContent = snap.content
        yield {
          type: 'init',
          content: lastContent,
        }
        continue
      }

      yield {
        type: 'new-snap',
        snap,
        index,
      }
      const steps = stepsTo(lastContent, snap.content)
      for (const step of steps)
        yield step

      lastContent = snap.content
    }
  }

  typewriter(options?: TypewriterOptions) {
    return typingAnimator(this.steps(), options)
  }
}

export type SnapshotFallbackLoader = (id: string) => Snapshots | undefined | Promise<Snapshots | undefined>
export interface SnapshotManagerOptions {
  ensureFallback?: SnapshotFallbackLoader
}

export class SnapshotManager extends Map<string, Snapshots> {
  constructor(
    public options: SnapshotManagerOptions = {},
  ) {
    super()
  }

  async ensure(
    id: string,
    load = this.options.ensureFallback,
  ) {
    if (!this.has(id))
      this.set(id, await load?.(id) || new Snapshots())
    return this.get(id)!
  }
}

export function getSnapshotPath(id: string) {
  if (id.endsWith(SNAP_EXT))
    return id
  return id + SNAP_EXT
}

export function getOriginalFilePath(id: string) {
  if (id.endsWith(SNAP_EXT))
    return id.slice(0, -SNAP_EXT.length)
  return undefined
}
