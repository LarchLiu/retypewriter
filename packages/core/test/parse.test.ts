import { expect, it } from 'vitest'
import { parseSnapshots } from '../src/state/parse'
import fixture from '../../../examples/main.js.retypewriter?raw'
import { Snapshots } from '../src'

it('works', () => {
  expect(parseSnapshots(fixture)).toMatchSnapshot()
})

it('serializable', () => {
  const snapshots = Snapshots.fromString(fixture)
  const stringified = snapshots.toString()
  expect(stringified).toEqual(fixture)
})
