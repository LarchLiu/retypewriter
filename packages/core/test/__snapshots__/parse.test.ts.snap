// Vitest Snapshot v1

exports[`works 1`] = `
{
  "head": {
    "options": {
      "delay": 100,
      "pause": true,
    },
  },
  "snapshots": [
    {
      "body": "",
      "bodyEnd": 65,
      "bodyStart": 65,
      "end": 65,
      "raw": "
--01----------
",
      "start": 49,
    },
    {
      "body": "import App from './App.vue'
",
      "bodyEnd": 109,
      "bodyStart": 81,
      "end": 109,
      "raw": "
--02----------
import App from './App.vue'
",
      "start": 65,
    },
    {
      "body": "import { } from 'vue'
import App from './App.vue'
",
      "bodyEnd": 175,
      "bodyStart": 125,
      "end": 175,
      "raw": "
--03----------
import { } from 'vue'
import App from './App.vue'
",
      "start": 109,
    },
    {
      "body": "import { createApp } from 'vue'
import App from './App.vue'

const app = createApp(App)
app.mount('#app')
",
      "bodyEnd": 297,
      "bodyStart": 191,
      "end": 324,
      "options": {
        "pause": true,
      },
      "optionsRaw": "pause: true",
      "raw": "
--04----------
import { createApp } from 'vue'
import App from './App.vue'

const app = createApp(App)
app.mount('#app')

-----options--
pause: true",
      "start": 175,
    },
    {
      "body": "import { createApp } from 'vue'
import { createRouter } from 'vue-router'
import App from './App.vue'

const app = createApp(App)
const router = createRouter()
app.use(router)
app.mount('#app')
",
      "bodyEnd": 534,
      "bodyStart": 340,
      "end": 534,
      "raw": "
--05----------
import { createApp } from 'vue'
import { createRouter } from 'vue-router'
import App from './App.vue'

const app = createApp(App)
const router = createRouter()
app.use(router)
app.mount('#app')
",
      "start": 324,
    },
    {
      "body": "import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import App from './App.vue'

const app = createApp(App)
const router = createRouter({
  history: createWebHistory(),
})
app.use(router)
app.mount('#app')
",
      "bodyEnd": 796,
      "bodyStart": 550,
      "end": 796,
      "raw": "
--06----------
import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import App from './App.vue'

const app = createApp(App)
const router = createRouter({
  history: createWebHistory(),
})
app.use(router)
app.mount('#app')
",
      "start": 534,
    },
    {
      "body": "import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import routes from 'virtual:generated-pages'
import App from './App.vue'

const app = createApp(App)
const router = createRouter({
  history: createWebHistory(),
  routes,
})
app.use(router)
app.mount('#app')
",
      "bodyEnd": 1113,
      "bodyStart": 812,
      "end": 1113,
      "raw": "
--07----------
import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import routes from 'virtual:generated-pages'
import App from './App.vue'

const app = createApp(App)
const router = createRouter({
  history: createWebHistory(),
  routes,
})
app.use(router)
app.mount('#app')
",
      "start": 796,
    },
  ],
}
`;
